vpr/SRC/route/route_tree_timing.c

```diff
@@ -479,16 +556,603 @@ void free_route_tree(t_rt_node * rt_node) {
+		rt_node->parent_switch = prev_edge->iswitch;
+		rt_node->parent_node = parent_node;
+		prev_edge->child = rt_node;
+
+		// at this point parent and child should be set and mutual
+		assert(parent_node->u.child_list->child == rt_node);
+		assert(rt_node->parent_node == parent_node);
+
+		rr_node_to_rt_node[tail->index] = rt_node;
+		// terminates this series of edges/branch
+		prev_edge = nullptr;
+
+		// either another prexisting part of the route tree and more sinks to come or end
+		head = tail->next;
+		// end of trunk
+		if (!head) return rt_root;
+		// continue on the next branch (1 after a sink for the traceback data structure)
+		// the rt_node has already been alocated as it's part of the trunk, make a new dangling edge for it
+		parent_node = rr_node_to_rt_node[head->index];
+		prev_edge = alloc_linked_rt_edge();
+		prev_edge->iswitch = head->iswitch;
+		// advance to trunk node's next edge
+		prev_edge->next = parent_node->u.child_list;
+		parent_node->u.child_list = prev_edge;
+		// edge is still dangling in preparation for next sink
+	}
+}
+
+static t_trace* traceback_branch_from_route_tree(t_trace* head, const t_rt_node* root, int& sinks_left) {
+
+	/* Transforms route tree (supposed to be a branch of the full tree) into
+	 * a traceback segement
+	 * returns the **sink trace (tail)** 
+	 * assumes head has been allocated with index set, but no iswitch or next */
+
+	t_linked_rt_edge* edge {root->u.child_list};
+	
+	for (;;) {
+		// head has been allocated at the start of this loop and has correct index
+
+		// reached a sink
+		if (!edge) {
+			head->iswitch = OPEN;
+			// inform caller that a sink was reached from there
+			--sinks_left;
+			// the caller will set the sink's next
+			return head;	// head is the sink
+		}
+		// join point (sibling edges)
+		else if (edge->next) {
+			// recursively update for each of its sibling branches
+			t_trace* sink;
+			do {
+				head->iswitch = edge->iswitch;
+				// a sub-branch off the current one, such that branch_head's parent is head
+				t_trace* branch_head = alloc_trace_data();
+				branch_head->index = edge->child->inode;
+				head->next = branch_head;
+
+				// the tip/end of the sub-branch
+				sink = traceback_branch_from_route_tree(branch_head, edge->child, sinks_left);
+				// danger here of losing a pointer if sink came from a branch point since sink->next would already be allocated
+				// don't allocate for last sink (no branches after the last sink)
+				if (sinks_left) {
+					sink->next = alloc_trace_data();	
+					// copy the information to the other instances (1 for each branch start) of the branch point
+					sink->next->index = head->index;
+					head = sink->next;	// become the since each branch might have different switches
+				}
+				edge = edge->next;
+			} while (edge);
+			return sink;	// returns the last ("rightmost") sink rooted at this branch point
+		}
+		// edge must be valid here, and there is only 1 edge
```

Comment:
For the following lines (713 - 720): might put them in "else {..}" to have a clearer structure. 
About the comment: might put "edge must be valid here" in separate line (or simply delete it), to show that "there is only 1 edge" is in comparison with "sink" & "joint point".